### DISCUSSION TITLE
Possibel approaches of automated builds for Windows

### DISCUSSION THREAD

**Discussion by Flyingmana** [tenacityteam/tenacity-legacy#183] [Ideas]
(trying to) extracting one of the sub topics in https://github.com/tenacityteam/tenacity/issues/155 to here

Question is, what ways are there to automate Builds for Windows (besides github actions)

**Comment by Be-ing** [comment]
GitLab.com has [Windows runners in beta](https://docs.gitlab.com/ee/ci/runners/build_cloud/windows_build_cloud.html).

**Comment by Be-ing** [comment]
AppVeyor supports Windows, but it is limited to a single job at a time on the free tier. I can say from experience that is inadequate for an actively maintained project.

**Reply by Be-ing** [reply]
What is so helpful about GitHub Actions is not just that it is cross platform, but the [computing resources available for free](https://docs.github.com/en/actions/reference/usage-limits-billing-and-administration#usage-limits) are very generous. It allows 20 concurrent jobs or 5 concurrent macOS jobs with a 6 hour time limit for jobs and no overall computing time limit. Any alternative would need some host to provide adequate computing power. Considering that Tenacity uses wxWidgets, I don't think we need 6 hour jobs (Mixxx does because the whole dependency graph takes about 4 hours to build, with about half of that time spent building Qt).

**Reply by falkTX** [reply]
This might just be the initial limits to get people to use it, then later on reduced once they get more people on it.
I do not think it is going to stay as-is for long.

**Reply by Be-ing** [reply]
That's a reasonable concern but I'm not too worried about that right now. I think GitHub has found a pretty decent solution to the issue of cryptocurrency mining abuse of CI by requiring manual approval before a contributor's first CI job gets run. So I don't think they're in any rush to limit these computing resources.

**Comment by Be-ing** [comment]
I have a proposal: use cross compiling for official builds while also supporting building on Windows and macOS. This could be maintained by moving development to another software forge where a CI service does the Linux builds and cross compiling from Linux. At the same time, we could keep GitHub as a read-only mirror and use GitHub Actions to ensure that building on Windows and macOS continues to work. This would allow developers on Windows and macOS to contribute code without the project needing to be tied to GitHub. However it might still be hard to fix platform-specific bugs in official builds. @falkTX what do you think about this?
